File: vane-regions/src/main/java/org/oddlama/vane/regions/event/RegionRoleSettingEnforcer.java
package org.oddlama.vane.regions.event;

import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.Tag;
import org.bukkit.block.Block;
import org.bukkit.block.Container;
import org.bukkit.block.DoubleChest;
import org.bukkit.entity.Entity;
import org.bukkit.entity.EntityType;
import org.bukkit.entity.ItemFrame;
import org.bukkit.entity.Minecart;
import org.bukkit.entity.Player;
import org.bukkit.entity.Projectile;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.block.BlockBreakEvent;
import org.bukkit.event.block.BlockPlaceEvent;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.event.hanging.HangingBreakByEntityEvent;
import org.bukkit.event.hanging.HangingPlaceEvent;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.event.inventory.InventoryDragEvent;
import org.bukkit.event.inventory.InventoryInteractEvent;
import org.bukkit.event.inventory.InventoryOpenEvent;
import org.bukkit.event.player.PlayerArmorStandManipulateEvent;
import org.bukkit.event.player.PlayerBucketEmptyEvent;
import org.bukkit.event.player.PlayerBucketFillEvent;
import org.bukkit.event.player.PlayerInteractEntityEvent;
import org.bukkit.event.player.PlayerInteractEvent;
import org.oddlama.vane.core.Listener;
import org.oddlama.vane.core.module.Context;
import org.oddlama.vane.regions.Regions;
import org.oddlama.vane.regions.region.RoleSetting;

public class RegionRoleSettingEnforcer extends Listener<Regions> {

    public RegionRoleSettingEnforcer(Context<Regions> context) {
        super(context);
    }

    public boolean checkSettingAt(
        final Location location,
        final Player player,
        final RoleSetting setting,
        final boolean checkAgainst
    ) {
        final var region = getModule().regionAt(location);
        if (region == null) {
            return false;
        }

        final var group = region.regionGroup(getModule());
        return group.getRole(player.getUniqueId()).getSetting(setting) == checkAgainst;
    }

    public boolean checkSettingAt(
        final Block block,
        final Player player,
        final RoleSetting setting,
        final boolean checkAgainst
    ) {
        final var region = getModule().regionAt(block);
        if (region == null) {
            return false;
        }

        final var group = region.regionGroup(getModule());
        return group.getRole(player.getUniqueId()).getSetting(setting) == checkAgainst;
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    public void onBlockBreak(final BlockBreakEvent event) {
        // Prevent breaking of region blocks
        if (checkSettingAt(event.getBlock(), event.getPlayer(), RoleSetting.BUILD, false)) {
            event.setCancelled(true);
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    public void onBlockPlace(final BlockPlaceEvent event) {
        // Prevent (re-)placing of region blocks
        if (checkSettingAt(event.getBlock(), event.getPlayer(), RoleSetting.BUILD, false)) {
            event.setCancelled(true);
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    public void onEntityDamageByEntity(final EntityDamageByEntityEvent event) {
        final var damaged = event.getEntity();
        final var damager = event.getDamager();

        switch (damaged.getType()) {
            default:
                return;
            case ARMOR_STAND: {
                if (!(damager instanceof Player)) {
                    break;
                }

                final var playerDamager = (Player) damager;
                if (checkSettingAt(damaged.getLocation().getBlock(), playerDamager, RoleSetting.BUILD, false)) {
                    event.setCancelled(true);
                }
                return;
            }
            case ITEM_FRAME: {
                if (!(damager instanceof Player)) {
                    break;
                }

                final var playerDamager = (Player) damager;
                final var itemFrame = (ItemFrame) damaged;
                final var item = itemFrame.getItem();
                if (item.getType() != Material.AIR) {
                    // This is a player taking the item out of an item-frame
                    if (
                        checkSettingAt(damaged.getLocation().getBlock(), playerDamager, RoleSetting.CONTAINER, false)
                    ) {
                        event.setCancelled(true);
                    }
                }
            }
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    public void onHangingBreakByEntity(final HangingBreakByEntityEvent event) {
        final Entity remover = event.getRemover();
        Player player = null;

        if (remover instanceof Player) {
            player = (Player) remover;
        } else if (remover instanceof Projectile) {
            final var projectile = (Projectile) remover;
            final var shooter = projectile.getShooter();
            if (shooter instanceof Player) {
                player = (Player) shooter;
            }
        }

        if (player != null && checkSettingAt(event.getEntity().getLocation(), player, RoleSetting.BUILD, false)) {
            event.setCancelled(true);
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    public void onHangingPlace(final HangingPlaceEvent event) {
        if (checkSettingAt(event.getEntity().getLocation(), event.getPlayer(), RoleSetting.BUILD, false)) {
            event.setCancelled(true);
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    public void onPlayerArmorStandManipulate(final PlayerArmorStandManipulateEvent event) {
        if (checkSettingAt(event.getRightClicked().getLocation(), event.getPlayer(), RoleSetting.BUILD, false)) {
            event.setCancelled(true);
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    public void onPlayerBucketEmpty(final PlayerBucketEmptyEvent event) {
        if (checkSettingAt(event.getBlockClicked(), event.getPlayer(), RoleSetting.BUILD, false)) {
            event.setCancelled(true);
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    public void onPlayerBucketFill(final PlayerBucketFillEvent event) {
        if (checkSettingAt(event.getBlockClicked(), event.getPlayer(), RoleSetting.BUILD, false)) {
            event.setCancelled(true);
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    public void onPlayerInteractEntity(final PlayerInteractEntityEvent event) {
        final var entity = event.getRightClicked();
        if (entity.getType() != EntityType.ITEM_FRAME) {
            return;
        }

        // Place or rotate item
        if (checkSettingAt(entity.getLocation(), event.getPlayer(), RoleSetting.CONTAINER, false)) {
            event.setCancelled(true);
        }
    }

    // The EventPriority is HIGH, so this is executed AFTER the portals try
    // to activate, which is a seperate permission.
    @EventHandler(priority = EventPriority.HIGH, ignoreCancelled = true)
    public void onPlayerInteract(final PlayerInteractEvent event) {
        final var player = event.getPlayer();
        final var block = event.getClickedBlock();
        if (block == null) {
            return;
        }

        switch (event.getAction()) {
            default:
                return;
            case PHYSICAL: {
                if (Tag.PRESSURE_PLATES.isTagged(block.getType())) {
                    if (checkSettingAt(block, player, RoleSetting.USE, false)) {
                        event.setCancelled(true);
                    }
                } else if (block.getType() == Material.TRIPWIRE) {
                    if (checkSettingAt(block, player, RoleSetting.USE, false)) {
                        event.setCancelled(true);
                    }
                }
                return;
            }
            case RIGHT_CLICK_BLOCK: {
                if (checkSettingAt(block, player, RoleSetting.USE, false)) {
                    event.setCancelled(true);
                }
            }
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    public void onPlayerInventoryOpen(final InventoryOpenEvent event) {
        // Only relevant if viewing should be prohibited, too.
        if (!getModule().configProhibitViewingContainers) {
            return;
        }

        if (!(event.getPlayer() instanceof Player player)) {
            return;
        }

        final var inventory = event.getInventory();
        if (inventory.getLocation() == null || inventory.getHolder() == null) {
            // Inventory is virtual / transient
            return;
        }

        final var holder = inventory.getHolder();
        if (holder instanceof DoubleChest || holder instanceof Container || holder instanceof Minecart) {
            if (checkSettingAt(inventory.getLocation(), player, RoleSetting.CONTAINER, false)) {
                event.setCancelled(true);
            }
        }
    }

    public void onPlayerInventoryInteract(final InventoryInteractEvent event) {
        final var clicker = event.getWhoClicked();
        if (!(clicker instanceof Player)) {
            return;
        }

        final var inventory = event.getInventory();
        if (inventory.getLocation() == null || inventory.getHolder() == null) {
            // Inventory is virtual / transient
            return;
        }

        final var holder = inventory.getHolder();
        if (holder instanceof DoubleChest || holder instanceof Container || holder instanceof Minecart) {
            if (checkSettingAt(inventory.getLocation(), (Player) clicker, RoleSetting.CONTAINER, false)) {
                event.setCancelled(true);
            }
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    public void onPlayerInventoryClick(final InventoryClickEvent event) {
        onPlayerInventoryInteract(event);
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    public void onPlayerInventoryDrag(final InventoryDragEvent event) {
        onPlayerInventoryInteract(event);
    }
}

File: vane-regions/src/main/kotlin/org/oddlama/vane/regions/event/RegionRoleSettingEnforcer.kt
package org.oddlama.vane.regions.event

import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.Tag
import org.bukkit.block.Block
import org.bukkit.block.Container
import org.bukkit.block.DoubleChest
import org.bukkit.entity.*
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.block.Action
import org.bukkit.event.block.BlockBreakEvent
import org.bukkit.event.block.BlockPlaceEvent
import org.bukkit.event.entity.EntityDamageByEntityEvent
import org.bukkit.event.hanging.HangingBreakByEntityEvent
import org.bukkit.event.hanging.HangingPlaceEvent
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.event.inventory.InventoryDragEvent
import org.bukkit.event.inventory.InventoryInteractEvent
import org.bukkit.event.inventory.InventoryOpenEvent
import org.bukkit.event.player.*
import org.oddlama.vane.core.Listener
import org.oddlama.vane.core.module.Context
import org.oddlama.vane.regions.Regions
import org.oddlama.vane.regions.region.Region
import org.oddlama.vane.regions.region.RoleSetting

class RegionRoleSettingEnforcer(context: Context<Regions?>?) : Listener<Regions?>(context) {
    fun checkSettingAt(
        location: Location,
        player: Player,
        setting: RoleSetting,
        checkAgainst: Boolean
    ): Boolean {
        val region: Region = module!!.regionAt(location) ?: return false

        val group = region.regionGroup(module!!)
        return group!!.getRole(player.uniqueId)!!.getSetting(setting) == checkAgainst
    }

    fun checkSettingAt(
        block: Block,
        player: Player,
        setting: RoleSetting,
        checkAgainst: Boolean
    ): Boolean {
        val region: Region = module!!.regionAt(block) ?: return false

        val group = region.regionGroup(module!!)
        return group!!.getRole(player.uniqueId)!!.getSetting(setting) == checkAgainst
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    fun onBlockBreak(event: BlockBreakEvent) {
        // Prevent breaking of region blocks
        if (checkSettingAt(event.getBlock(), event.player, RoleSetting.BUILD, false)) {
            event.isCancelled = true
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    fun onBlockPlace(event: BlockPlaceEvent) {
        // Prevent (re-)placing of region blocks
        if (checkSettingAt(event.getBlock(), event.getPlayer(), RoleSetting.BUILD, false)) {
            event.isCancelled = true
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    fun onEntityDamageByEntity(event: EntityDamageByEntityEvent) {
        val damaged = event.getEntity()
        val damager = event.damager

        when (damaged.type) {
            EntityType.ARMOR_STAND -> {
                if (damager !is Player) {
                    return
                }

                if (checkSettingAt(damaged.location.block, damager, RoleSetting.BUILD, false)) {
                    event.isCancelled = true
                }
                return
            }

            EntityType.ITEM_FRAME -> {
                if (damager !is Player) {
                    return
                }

                val itemFrame = damaged as ItemFrame
                val item = itemFrame.item
                if (item.type != Material.AIR) {
                    // This is a player taking the item out of an item-frame
                    if (checkSettingAt(damaged.location.block, damager, RoleSetting.CONTAINER, false)
                    ) {
                        event.isCancelled = true
                    }
                }
            }

            else -> return
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    fun onHangingBreakByEntity(event: HangingBreakByEntityEvent) {
        val remover = event.remover
        var player: Player? = null

        if (remover is Player) {
            player = remover
        } else if (remover is Projectile) {
            val shooter = remover.shooter
            if (shooter is Player) {
                player = shooter
            }
        }

        if (player != null && checkSettingAt(event.entity.location, player, RoleSetting.BUILD, false)) {
            event.isCancelled = true
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    fun onHangingPlace(event: HangingPlaceEvent) {
        if (checkSettingAt(event.entity.location, event.player!!, RoleSetting.BUILD, false)) {
            event.isCancelled = true
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    fun onPlayerArmorStandManipulate(event: PlayerArmorStandManipulateEvent) {
        if (checkSettingAt(event.rightClicked.location, event.getPlayer(), RoleSetting.BUILD, false)) {
            event.isCancelled = true
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    fun onPlayerBucketEmpty(event: PlayerBucketEmptyEvent) {
        if (checkSettingAt(event.blockClicked, event.getPlayer(), RoleSetting.BUILD, false)) {
            event.isCancelled = true
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    fun onPlayerBucketFill(event: PlayerBucketFillEvent) {
        if (checkSettingAt(event.blockClicked, event.getPlayer(), RoleSetting.BUILD, false)) {
            event.isCancelled = true
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    fun onPlayerInteractEntity(event: PlayerInteractEntityEvent) {
        val entity = event.rightClicked
        if (entity.type != EntityType.ITEM_FRAME) {
            return
        }

        // Place or rotate item
        if (checkSettingAt(entity.location, event.getPlayer(), RoleSetting.CONTAINER, false)) {
            event.isCancelled = true
        }
    }

    // The EventPriority is HIGH, so this is executed AFTER the portals try
    // to activate, which is a seperate permission.
    @EventHandler(priority = EventPriority.HIGH, ignoreCancelled = true)
    fun onPlayerInteract(event: PlayerInteractEvent) {
        val player = event.getPlayer()
        val block = event.clickedBlock ?: return

        when (event.action) {
            Action.PHYSICAL -> {
                if (Tag.PRESSURE_PLATES.isTagged(block.type)) {
                    if (checkSettingAt(block, player, RoleSetting.USE, false)) {
                        event.isCancelled = true
                    }
                } else if (block.type == Material.TRIPWIRE) {
                    if (checkSettingAt(block, player, RoleSetting.USE, false)) {
                        event.isCancelled = true
                    }
                }
                return
            }

            Action.RIGHT_CLICK_BLOCK -> {
                if (checkSettingAt(block, player, RoleSetting.USE, false)) {
                    event.isCancelled = true
                }
            }

            else -> return
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    fun onPlayerInventoryOpen(event: InventoryOpenEvent) {
        // Only relevant if viewing should be prohibited, too.
        if (!module!!.configProhibitViewingContainers) {
            return
        }

        if (event.player !is Player) {
            return
        }

        val player = event.player as Player
        if (checkContainerSettingAt(event.inventory, player)) {
            event.isCancelled = true
        }
    }

    private fun checkContainerSettingAt(inventory: org.bukkit.inventory.Inventory, player: Player): Boolean {
        if (inventory.location == null || inventory.holder == null) {
            // Inventory is virtual / transient
            return false
        }

        val holder = inventory.holder
        return (holder is DoubleChest || holder is Container || holder is Minecart)
            && checkSettingAt(inventory.location!!, player, RoleSetting.CONTAINER, false)
    }

    fun onPlayerInventoryInteract(event: InventoryInteractEvent) {
        val clicker = event.whoClicked
        if (clicker !is Player) {
            return
        }

        if (checkContainerSettingAt(event.inventory, clicker)) {
            event.isCancelled = true
        }
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    fun onPlayerInventoryClick(event: InventoryClickEvent) {
        onPlayerInventoryInteract(event)
    }

    @EventHandler(priority = EventPriority.LOW, ignoreCancelled = true)
    fun onPlayerInventoryDrag(event: InventoryDragEvent) {
        onPlayerInventoryInteract(event)
    }
}
